src/scheduler/Scheduler.cpp

```diff
@@ -549,6 +582,29 @@ FunctionCallClient& Scheduler::getFunctionCallClient(
     return it->second;
 }
 
+SnapshotClient& Scheduler::getSnapshotClient(const std::string& otherHost)
+{
+    // Note, our keys here have to include the tid as the clients can only be
+    // used within the same thread
```

Comment:
@csegarragonz I don't know how this isn't failing for the function call clients, as the `functionCallClients` map is shared by all threads and keyed on the other hostname, so I would imagine the same client is getting used by different threads. However, it seems there aren't any assertions failing, so I guess that isn't true...?

Reply:
It is not failing as we do the assertions, but I'm indeed surprised that it is not. Is your approach thread-safe either? I see that different threads would only ever access different keys, but I am wondering if this won't leave the unerlying data structure in an undefined state.

I guess I assumed the scheduler was single-threaded. In MPI we use TLS for the clients, but it's painful to clean up (as it must be explicitely cleaned, from the same thread which opened it). Happy to discuss this offline.

Reply:
The scheduler is not single-threaded, it's a singleton object that gets called by lots of threads, so thread safety is still very important. The confusion may have arisen because the scheduler used to manage its own thread pool but now doesn't actually spawn any threads (the executors do that), this does not mean it's only ever run in a single thread though.

You're right on the map accesses, there should be a lock around this as the map itself isn't thread safe.